examples/pde.hpp

```diff
@@ -342,6 +342,141 @@ double HalfCoeffecient::Eval(mfem::ElementTransformation& T,
     return transip(1) < (spe10_scale_ * 44 * 5) ? -value_ : value_;
 }
 
+/**
+ @brief compute transmissibility based on two-point flux approximation
+
+ The implementation is based on (Sec. 4.4.1 of) the preliminary version of
+
+ K.-A. Lie. An introduction to reservoir simulation using MATLAB/GNU Octave
+ **/
+class LocalTPFA
+{
+    mfem::Coefficient* Q_;
+    mfem::VectorCoefficient* VQ_;
+    const mfem::ParMesh& mesh_;
+    const mfem::SparseMatrix& vert_edge_;
+
+    mfem::DenseMatrix EvalKappa(int i);
+    mfem::Vector ComputeShapeCenter(const mfem::Element& el);
+    mfem::Vector ComputeLocalWeight(int i);
+public:
+    LocalTPFA(const mfem::ParMesh& mesh, const mfem::SparseMatrix& vert_edge)
+        : Q_(NULL), VQ_(NULL), mesh_(mesh), vert_edge_(vert_edge) { }
+
+    /// @param q inverse of permeability \f$ kappa^{-1} \f$ in Darcy's law
+    LocalTPFA(const mfem::ParMesh& mesh, const mfem::SparseMatrix& vert_edge,
+              mfem::Coefficient& q)
+        : Q_(&q), VQ_(NULL), mesh_(mesh), vert_edge_(vert_edge) { }
+
+    LocalTPFA(const mfem::ParMesh& mesh, const mfem::SparseMatrix& vert_edge,
+              mfem::VectorCoefficient& q)
+        : Q_(NULL), VQ_(&q), mesh_(mesh), vert_edge_(vert_edge) { }
+
+    /// Compute local edge weights for the corresponding graph Laplacian
+    std::vector<mfem::Vector> ComputeLocalWeights()
+    {
+        std::vector<mfem::Vector> local_weights(mesh_.GetNE());
+        for (int i = 0; i < mesh_.GetNE(); ++i)
+        {
+            local_weights[i] = ComputeLocalWeight(i);
+        }
+        return local_weights;
+    }
+};
+
+mfem::DenseMatrix LocalTPFA::EvalKappa(int i)
+{
+    const int dim = mesh_.Dimension();
+    const mfem::Element& el = *(mesh_.GetElement(i));
+    auto trans = const_cast<mfem::ParMesh&>(mesh_).GetElementTransformation(i);
+    const auto& ip = mfem::IntRules.Get(el.GetType(), 1).IntPoint(0);
+
+    mfem::DenseMatrix kappa(dim);
+
+    // Note that Q is kappa^{-1}
+    if (VQ_)
+    {
+        mfem::Vector vq(dim);
+        VQ_->Eval(vq, *trans, ip);
+        for (int d = 0; d < dim; ++d)
+            kappa(d, d) = 1.0 / vq(d);
+    }
+    else if (Q_)
+    {
+        double sq = Q_->Eval(*trans, ip);
+        for (int d = 0; d < dim; ++d)
+            kappa(d, d) = 1.0 / sq;
+    }
+    else
+    {
+        for (int d = 0; d < dim; ++d)
+            kappa(d, d) = 1.0;
+    }
+
+    return kappa;
+}
+
+mfem::Vector LocalTPFA::ComputeShapeCenter(const mfem::Element& el)
```

Comment:
I think it will be simpler (and more general) to call `mfem::Mesh::GetElementCenter()`).

Reply:
Yes, this is a better way. I was not aware of this function, thanks for pointing it out. By the way, do we stick with `mfem v3.3` (as in the `INTALL.md`) or just the latest version? Since I couldn't find `Mesh::GetElementCenter()` in `v3.3.2` (the one that I am using)?

Reply:
That's a good point. I'm apparently using a newer MFEM (which works fine) but I'm not sure this particular thing is a good reason to upgrade the whole project. Either way is fine with me.

Reply:
Is there some similar function for computing face/edge center as well? If not, we have to keep this extra function because face/edge center are needed as well.

Reply:
I don't see one, so maybe we should just leave it.